contracts/core/OptionsPremiumPricer.sol

```diff
@@ -0,0 +1,227 @@
+            _c = max(_p.add(sp), st) == st ? 0 : _p.add(sp).sub(st);
+        }
+
+        return (_c, _p);
+    }
+
+    /**
+     * @notice Calculates black scholes for the ITM option at mint given strike
+     * price and underlying given the parameters (if underling >= strike price this is
+     * premium of call, and put otherwise)
+     * @param t is the days until expiry
+     * @param v is the annualized volatility
+     * @param sp is the underlying price
+     * @param st is the strike price
+     * @return premium is the premium of option
+     */
+    function C(
+        uint256 t,
+        uint256 v,
+        uint256 sp,
+        uint256 st
+    ) private view returns (uint256 premium) {
+        if (sp == st) {
+            return
+                (((((Math.LNX * sp) / 1e10) * v) / 1e18) *
+                    Math.sqrt2((1e18 * t) / 365)) / 1e9;
+        }
+
+        (uint256 d1, uint256 d2) = d(t, v, sp, st);
+
+        uint256 cdfD1 = Math.ncdf((Math.FIXED_1 * d1) / 1e18);
+        uint256 cdfD2 = Math.cdf((int256(Math.FIXED_1) * int256(d2)) / 1e18);
+
+        premium = (sp * cdfD1) / 1e14 - (st * cdfD2) / 1e14;
+    }
+
+    /**
+     * @notice Calculates d1 and d2 used in black scholes calculation
+     * as parameters to black scholes calculations
+     * @param t is the days until expiry
+     * @param v is the annualized volatility
+     * @param sp is the underlying price
+     * @param st is the strike price
+     * @return d1 and d2
+     */
+    function d(
+        uint256 t,
+        uint256 v,
+        uint256 sp,
+        uint256 st
+    ) private view returns (uint256 d1, uint256 d2) {
+        uint256 sigma = ((v**2) / 2);
+        uint256 sigmaB = 1e36;
+
+        uint256 sig = (((1e18 * sigma) / sigmaB) * t) / 365;
+
+        uint256 sSQRT = (v * Math.sqrt2((1e18 * t) / 365)) / 1e9;
+
+        d1 = (1e18 * Math.ln((Math.FIXED_1 * sp) / st)) / Math.FIXED_1;
+        d1 = ((d1 + sig) * 1e18) / sSQRT;
+        d2 = d1 - sSQRT;
+    }
+
+    /**
+     * @notice Calculates the current underlying price, annualized volatility, and days until expiry
+     * as parameters to black scholes calculations
+     * @param expiryTimestamp is the unix timestamp of expiry
+     * @return sp is the underlying
+     * @return v is the volatility
+     * @return t is the days until expiry
+     */
+    function blackScholesParams(uint256 expiryTimestamp)
+        private
+        view
+        returns (
+            uint256 sp,
+            uint256 v,
+            uint256 t
+        )
+    {
+        // chainlink oracle returns crypto / usd pairs with 8 decimals, like otoken strike price
+        sp = priceOracle.latestAnswer().div(10**priceOracle.decimals()).mul(
+            10**8
+        );
+        // annualized vol * 10 ** 7 because delta expects 16 decimals
+        // and annualizedVol is 8 decimals + it shifts percentage by one decimal place
+        v = volatilityOracle.annualizedVol().mul(10**7);
```

Comment:
there is some decimal weirdness wrt volatility. apparently shifting it to only 16 decimals aligns it with online calculators (ex: 1.2% vol actually resolves to 120% vol so I just multiplied by 10 ** 7 instead of 10 ** 9)

good resources to double check my black scholes:

https://goodcalculators.com/black-scholes-calculator/
https://www.calkoo.com/en/black-scholes-option-pricing-model